### PR TITLE
[inductor] Remove comm-specific node attributes from scheduler

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129346
* #129344
* #129343
* #129325
* #128893
* #128979
* __->__ #129084

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -1901,9 +1856,6 @@ def replace_buffer(
                 new_scheduler_node.min_order = node.min_order
                 new_scheduler_node.max_order = node.max_order
                 new_scheduler_node.last_usage = node.last_usage
-                for user in new_scheduler_node.users:
-                    user.node.inverse_users.remove(node)
-                    user.node.inverse_users.append(new_scheduler_node)
```
Comment (lezcano) [inline torch/_inductor/scheduler.py:1906]: where is this done now? Is it recomputed every time is needed?

**Review by lezcano — COMMENTED** [review]
I worry that the tests have not caught that bug. Where is this code exercised?
Also, is this now recomputing everything every time it needs it? If so, we should probably benchmark and consider using a cache.

File: torch/_inductor/comms.py [hunk 2]
```diff
@@ -18,6 +19,8 @@
 
 def sink_waits(
     snodes: List[BaseSchedulerNode],
+    node_users: Dict[BaseSchedulerNode, Set[BaseSchedulerNode]],
+    inverse_users: Dict[BaseSchedulerNode, Set[BaseSchedulerNode]],
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/comms.py:23]: On line 410, all of these comms passes are called in a loop with the same signature so they have to match.

File: torch/_inductor/comms.py [hunk 3]
```diff
@@ -56,10 +61,8 @@ def raise_comms(
         if is_collective(snode.node):
             cur_comms.append(snode)
         else:
-            for comm in cur_comms:
-                assert len(comm.inverse_users) > 0
             while len(cur_comms) > 0 and any(
-                snode in comm.inverse_users for comm in cur_comms
+                comm in node_users[snode] for comm in cur_comms
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/comms.py:65]: Note that I swapped the order here, so `A in node_users[B]` <=> `B in inverse_users[A]`. There isn't really any point now though, so I guess I can revert.

File: torch/_inductor/comms.py [hunk 4]
```diff
@@ -40,6 +43,8 @@ def sink_waits(
 
 def raise_comms(
     snodes: List[BaseSchedulerNode],
+    node_users: Dict[BaseSchedulerNode, Set[BaseSchedulerNode]],
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply torch/_inductor/comms.py:46]: As above, the signature needs to match the other passes.

File: torch/_inductor/scheduler.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/scheduler.py:1906]: The comms passes are only called once so this was pointless book-keeping. We compute it once and discard it when the pass is finished, so don't need to keep it up to date.

**Review by lezcano — APPROVED** [review]

**Pull request closed without merging**